### PR TITLE
Expanded  grpc calls

### PR DESCRIPTION
**Pull request by SWvheerden**
## Description

Expanded the basenode grpc server to include the following calls:
Add get_block_template
Add get_block
Add get_meta_data
Add submit_block

## Motivation and Context

This allows a miner to connect the base_node via grpc. 

## Types of changes

* [ ] Bug fix (non-breaking change which fixes an issue)
* [x] New feature (non-breaking change which adds functionality)
* [ ] Breaking change (fix or feature that would cause existing functionality to change)
* [ ] Feature refactor (No new feature or functional changes, but performance or technical debt improvements)
* [ ] New Tests
* [ ] Documentation

## Checklist:

* [x] I'm merging against the `development` branch.
* [x] I ran `cargo-fmt --all` before pushing.
* [x] I have squashed my commits into a single commit.
* [ ] My change requires a change to the documentation.
* [ ] I have updated the documentation accordingly.
* [ ] I have added tests to cover my changes.

### CODE REVIEW

File: applications/tari_base_node/src/grpc/server.rs [hunk 1]
```diff
@@ -296,6 +314,76 @@ impl base_node_grpc::base_node_server::BaseNode for BaseNodeGrpcServer {
...
+        debug!(target: LOG_TARGET, "Incoming GRPC request for get new block template");
+        let algo = match request.pow_algo {
+            0 => PowAlgorithm::Monero,
+            1 => PowAlgorithm::Blake,
+            _ => return Err(Status::invalid_argument("No valid pow algo selected".to_string())),
+        };
+        let mut handler = self.node_service.clone();
+
+        let new_template = handler
+            .get_new_block_template(PowAlgorithm::Blake)
```
Comment (philipr-za) [inline applications/tari_base_node/src/grpc/server.rs]: Shouldn't this be `algo`?

**Review by StriderDM — DISMISSED** [review]
Thanks for this! 

LGTM

File: applications/tari_base_node/src/grpc/server.rs [hunk 2]
```diff
@@ -666,7 +778,272 @@ impl From<BlockHeader> for base_node_grpc::BlockHeader {
...
+impl From<NewBlockTemplate> for base_node_grpc::NewBlockTemplate {
+    fn from(block: NewBlockTemplate) -> Self {
+        let header = base_node_grpc::NewBlockHeaderTemplate {
+            version: block.header.version as u32,
+            height: block.header.height,
+            prev_hash: block.header.prev_hash.clone(),
+            total_kernel_offset: Vec::from(block.header.total_kernel_offset.as_bytes()),
+            pow: Some(base_node_grpc::ProofOfWork {
+                pow_algo: match block.header.pow.pow_algo {
+                    PowAlgorithm::Monero => 0,
```
Comment (philipr-za) [inline applications/tari_base_node/src/grpc/server.rs]: I think you should rather implement a `From<PowAlgorithm> for u32` and a `TryFrom<u32> for PowAlgorithm` to handle these conversions. That way it is maintained in 1 clear place.

File: applications/tari_base_node/src/grpc/server.rs [hunk 3]
```diff
@@ -296,6 +314,76 @@ impl base_node_grpc::base_node_server::BaseNode for BaseNodeGrpcServer {
...
     }
 
+    async fn get_new_block_template(
+        &self,
+        request: Request<base_node_grpc::PowAlgo>,
+    ) -> Result<Response<base_node_grpc::NewBlockTemplate>, Status>
+    {
+        let request = request.into_inner();
+        debug!(target: LOG_TARGET, "Incoming GRPC request for get new block template");
+        let algo = match request.pow_algo {
```
Comment (philipr-za) [inline applications/tari_base_node/src/grpc/server.rs]: I think you should rather implement a `From<PowAlgorithm> for u32` and a `TryFrom<u32> for PowAlgorithm` to handle these conversions. That way it is maintained in 1 clear place.

File: applications/tari_base_node/proto/base_node.proto [hunk 4]
```diff
@@ -223,6 +231,14 @@ message ProofOfWork {
...
     uint64 accumulated_blake_difficulty = 3;
     bytes pow_data = 4;
+    uint64 target_difficulty = 5;
+}
+
+//This is used to request the which pow algo should be used with the block template
+message PowAlgo{
+    // 0 = Monero
+    // 1 = Blake
+    uint64 pow_algo = 1;
```
Comment (philipr-za) [inline applications/tari_base_node/proto/base_node.proto]: Does this need to be a u64?

Comment (SWvheerden) [reply applications/tari_base_node/proto/base_node.proto]: made smaller

File: applications/tari_base_node/src/grpc/server.rs (hunk 3 above)
Comment (SWvheerden) [reply applications/tari_base_node/src/grpc/server.rs]: We have one, used that one

File: applications/tari_base_node/src/grpc/server.rs (hunk 1 above)
Comment (SWvheerden) [reply applications/tari_base_node/src/grpc/server.rs]: yip, thanks, fixed

File: applications/tari_base_node/src/grpc/server.rs (hunk 2 above)
Comment (SWvheerden) [reply applications/tari_base_node/src/grpc/server.rs]: Wel thats currently used in a from, that means we have to change the entire conversion here to a try_from. 
At this point we dont have to check account for try's.

File: applications/tari_base_node/src/grpc/server.rs [hunk 5]
```diff
@@ -666,7 +776,272 @@ impl From<BlockHeader> for base_node_grpc::BlockHeader {
...
+        Self {
+            height_of_longest_chain: meta.height_of_longest_chain.unwrap_or(0),
+            best_block: meta.best_block.unwrap_or(vec![]),
+            pruning_horizon: meta.pruning_horizon,
+            accumulated_difficulty: diff.as_u64(),
+        }
+    }
+}
+
+impl From<NewBlockTemplate> for base_node_grpc::NewBlockTemplate {
```
Comment (stringhandler) [inline applications/tari_base_node/src/grpc/server.rs]: I think put these conversions in a new file, or even separate files. This file is getting pretty big and these may be used elsewhere

**Review by stringhandler — DISMISSED** [review]
Looks good. I think it'd be neater with the conversions moved into their own files, if you have time, but if not we can merge it as is.

**Review by stringhandler — APPROVED** [review]

**Pull request merged**